File: sol/solution/0101-0200/0106/Solution.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func buildTree(inorder []int, postorder []int) *TreeNode {
    idx := make(map[int]int, len(inorder))
    for i, v := range inorder { idx[v] = i }
    var build func(il, ir int) *TreeNode
    build = func(il, ir int) *TreeNode {
        if il > ir { return nil }
        val := postorder[len(postorder)-1]
        postorder = postorder[:len(postorder)-1]
        node := &TreeNode{Val: val}
        split := idx[val]
        node.Right = build(split+1, ir)
        node.Left  = build(il, split-1)
        return node
    }
    return build(0, len(inorder)-1)
}

File: sol/solution/0101-0200/0106/Solution.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def buildTree(self, inorder: List[int], postorder: List[int]) -> Optional[TreeNode]:
        idx = {v: i for i, v in enumerate(inorder)}
        def build(l, r):
            if l > r: return None
            root = TreeNode(postorder.pop())
            i = idx[root.val]
            root.right = build(i + 1, r)
            root.left = build(l, i - 1)
            return root
        return build(0, len(inorder) - 1)